parsers/US_PREPA.py

```diff
@@ -14,59 +14,227 @@
 from logging import Logger, getLogger
 from typing import Optional
 
-# The arrow library is used to handle datetimes
 import arrow
-from requests import Session
+from requests import Response, Session
+
+from parsers.lib.exceptions import ParserException
 
 timezone_name = "America/Puerto_Rico"
 US_PROXY = "https://us-ca-proxy-jfnx5klx2a-uw.a.run.app"
 HOST_PARAMETER = "?host=https://aeepr.com"
+
 GENERATION_BREAKDOWN_URL = (
-    f"{US_PROXY}/es-pr/generacion/Documents/combustibles.aspx{HOST_PARAMETER}"
-)
-RENEWABLES_BREAKDOWN_URL = (
-    f"{US_PROXY}/es-pr/generacion/Documents/Unidades_renovables.aspx{HOST_PARAMETER}"
-)
-TIMESTAMP_URL = (
-    f"{US_PROXY}/es-pr/generacion/Documents/CostosCombustible.aspx{HOST_PARAMETER}"
+    f"https://aeepr.com/es-pr/generacion/Documents/DataJS/dataSource.js"
+    # f"{US_PROXY}/es-pr/generacion/Documents/DataJS/dataSource.js{HOST_PARAMETER}"
 )
 
+"""
+Two more endpoints, not needed here. Contain status of power plants, and 24h history of frequency and generation sum.
+/es-pr/generacion/Documents/DataJS/dataGraph.js"
+/es-pr/generacion/Documents/DataJS/dataUnitStatus.js"
+"""
 
-def extract_data(html):
-    """Extracts data from the source code of an HTML page with a FusionCharts chart"""
-    dataSource = re.search(r"dataSource: (\{.+\}\]\})\}\);", html).group(
-        1
-    )  # Extract object with data
-    dataSource = re.sub(
-        r",\s*\}", "}", dataSource
-    )  # ,} is valid JavaScript but not valid JSON
-    dataSource = json.loads(dataSource)
-    sourceData = dataSource[
-        "data"
-    ]  # The rest of the dataSource object contains unnecessary stuff like chart theme, label, axis names etc.
-    return sourceData
+VAILDATION_THRESHOLD_PERCENT = 2.0
+
+# Source data identifiers -> Production types. See also comments at oil/gas validation
+SRC_POWER_TYPE_TO_PRODUCTION_TYPE = {
+    # fossil
+    "Turbina de Gas": "gas",
```

Comment:
As I pointed out in my parser, in spite of the name, there are gas turbines that run on diesel; 'gas' doesn't mean 'natural gas' here

Reply:
But if you've found a data source that indicates all Puerto Rican gas turbines use natural gas, I'm interested to know

Reply:
Nope, I also can't provide a specific source for this specific information. I also found this the hard way first by assuming first they are based on gas, indeed. That's why I labeled them first as "gas" there. 

However, [later](https://github.com/MuffinCompiler/electricitymaps-contrib-pr/blob/4b27e41cbb6c96aec554975ea08903b84e58dc23/parsers/US_PREPA.py#L332-L359), the actual amounts of oil/gas are infered from the ratio source table in the data source, and not accumulated over each generation site. Therefore, the returned production values for 'gas' and 'oil' should be most accurate. Probably I should have set the type as "combined" instead of "gas" to express this.

Reply:
I just simplified the parser to infer coal, gas, and oil only from the % table. Renewables are parsed more specifically by type.